### 1. Predecir el valor del dolar utilizando una red neuronal

Resolver el problema de predecir el valor del dólar, que se planteó en el capítulo
anterior, utilizando una red neuronal. Describir las diferencias entre un modelo de
entrenamiento y el otro

In [2]:
%pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

def limpiar_y_convertir_a_float(df, columnas):
    for columna in columnas:
        df[columna] = df[columna].str.replace('.', '').str.replace(',', '.').astype(float)
    return df

bancolombia_df = pd.read_csv('Datos históricos BIC (1).csv')
ecopetrol_df = pd.read_csv('Datos históricos ECO (2).csv')
petroleo_df = pd.read_csv('Datos históricos Futuros petróleo Brent.csv')
tasa_cambio_df = pd.read_csv('Tasa_de_Cambio_Representativa_del__Mercado_-Historico_20240404 (1).csv')

columnas_para_convertir = ['Último', 'Apertura', 'Máximo', 'Mínimo'] 

bancolombia_df = limpiar_y_convertir_a_float(bancolombia_df, columnas_para_convertir)
ecopetrol_df = limpiar_y_convertir_a_float(ecopetrol_df, columnas_para_convertir)
petroleo_df = limpiar_y_convertir_a_float(petroleo_df, columnas_para_convertir)

bancolombia_df['Fecha'] = pd.to_datetime(bancolombia_df['Fecha'], format='%d.%m.%Y').dt.strftime('%Y-%m-%d')
ecopetrol_df['Fecha'] = pd.to_datetime(ecopetrol_df['Fecha'], format='%d.%m.%Y').dt.strftime('%Y-%m-%d')
petroleo_df['Fecha'] = pd.to_datetime(petroleo_df['Fecha'], format='%d.%m.%Y').dt.strftime('%Y-%m-%d')
tasa_cambio_df['Fecha'] = pd.to_datetime(tasa_cambio_df['VIGENCIADESDE'], format='%d/%m/%Y').dt.strftime('%Y-%m-%d')

bancolombia_precios = bancolombia_df[['Fecha', 'Último']].rename(columns={'Último': 'Precio_Bancolombia'})
ecopetrol_precios = ecopetrol_df[['Fecha', 'Último']].rename(columns={'Último': 'Precio_Ecopetrol'})
petroleo_precios = petroleo_df[['Fecha', 'Último']].rename(columns={'Último': 'Precio_Petroleo'})
tasa_cambio = tasa_cambio_df[['Fecha', 'VALOR']].rename(columns={'VALOR': 'Tasa_Cambio'})

datos_combinados = pd.merge(bancolombia_precios, ecopetrol_precios, on='Fecha', how='outer')
datos_combinados = pd.merge(datos_combinados, petroleo_precios, on='Fecha', how='outer')
datos_combinados = pd.merge(datos_combinados, tasa_cambio, on='Fecha', how='outer')

datos_combinados = datos_combinados.sort_values('Fecha')



datos_completos = datos_combinados.interpolate().dropna()

X = datos_completos[['Precio_Bancolombia', 'Precio_Ecopetrol', 'Precio_Petroleo']]
y = datos_completos['Tasa_Cambio']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

nn_model = make_pipeline(StandardScaler(), MLPRegressor(hidden_layer_sizes=(100, 50), activation='relu', random_state=42, max_iter=500))
nn_model.fit(X_train, y_train)

y_pred = nn_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
rmse = np.sqrt(mse)

print(f'MSE: {mse}, R^2: {r2}, RMSE: {rmse}')


C:\Users\dmriv\AppData\Local\Temp\ipykernel_2984\179312154.py:48: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  datos_completos = datos_combinados.interpolate().dropna()


MSE: 76677.19990005113, R^2: 0.8954941716314369, RMSE: 276.9064822282987


c:\Users\dmriv\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


MSE (Error Cuadrático Medio): 76,677.20. Un valor más bajo indica un mejor ajuste.

R² (Coeficiente de Determinación): 0.895. Esta es una métrica clave que indica cuánta variabilidad de tus datos puede ser explicada por el modelo.valor de 0.895 sugiere que el modelo explica aproximadamente el 89.5% de la variabilidad, lo cual es bastante bueno.

RMSE (Raíz del Error Cuadrático Medio): 276.91. Esta métrica ofrece una medida de la magnitud del error, proporcionando una idea de cuán grandes son los errores en términos de las unidades originales de los datos

In [2]:
pip install tensorflow

  Using cached requests-2.31.0-py3-none-any.whl.metadata (4.6 kB)
  Using cached rich-13.7.1-py3-none-any.whl.metadata (18 kB)
     ---------------------------------------- 0.0/46.2 kB ? eta -:--:--
     ---------------------------------------- 46.2/46.2 kB ? eta 0:00:00
  Using cached charset_normalizer-3.3.2-cp312-cp312-win_amd64.whl.metadata (34 kB)
  Using cached idna-3.6-py3-none-any.whl.metadata (9.9 kB)
  Using cached urllib3-2.2.1-py3-none-any.whl.metadata (6.4 kB)
  Using cached certifi-2024.2.2-py3-none-any.whl.metadata (2.2 kB)
  Using cached markdown_it_py-3.0.0-py3-none-any.whl.metadata (6.9 kB)
  Using cached mdurl-0.1.2-py3-none-any.whl.metadata (1.6 kB)
   ---------------------------------------- 0.0/377.1 MB ? eta -:--:--
   ---------------------------------------- 0.6/377.1 MB 11.5 MB/s eta 0:00:33
   ---------------------------------------- 1.0/377.1 MB 11.0 MB/s eta 0:00:35
   ---------------------------------------- 1.2/377.1 MB 8.6 MB/s eta 0:00:44
   ------------

In [1]:
pip install keras

Note: you may need to restart the kernel to use updated packages.


In [7]:
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator

train_data_dir = 'C:/Users/dmriv/Documents/GitHub/IAminibots/fruits-360/Training'
test_data_dir = 'C:/Users/dmriv/Documents/GitHub/IAminibots/fruits-360/Test'

model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(150, 150, 3)),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(128, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(128, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation='relu'),
    tf.keras.layers.Dense(27, activation='softmax')
])

model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
    train_data_dir,
    target_size=(150, 150),
    batch_size=32,
    class_mode='categorical')

test_generator = test_datagen.flow_from_directory(
    test_data_dir,
    target_size=(150, 150),
    batch_size=32,
    class_mode='categorical')

model.fit(
    train_generator,
    steps_per_epoch=train_generator.samples // train_generator.batch_size,
    epochs=10,
    validation_data=test_generator,
    validation_steps=test_generator.samples // test_generator.batch_size)

Found 14101 images belonging to 27 classes.
Found 4717 images belonging to 27 classes.
Epoch 1/10


c:\Users\dmriv\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:120: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


440/440 ━━━━━━━━━━━━━━━━━━━━ 142s 319ms/step - accuracy: 0.7259 - loss: 0.9157 - val_accuracy: 0.9071 - val_loss: 0.3061
Epoch 2/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 0s 139us/step - accuracy: 0.9688 - loss: 0.1252 - val_accuracy: 0.9231 - val_loss: 0.2052
Epoch 3/10


c:\Users\dmriv\AppData\Local\Programs\Python\Python312\Lib\contextlib.py:158: UserWarning: Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches. You may need to use the `.repeat()` function when building your dataset.
  self.gen.throw(value)


440/440 ━━━━━━━━━━━━━━━━━━━━ 93s 211ms/step - accuracy: 0.9931 - loss: 0.0245 - val_accuracy: 0.9354 - val_loss: 0.2083
Epoch 4/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 0s 85us/step - accuracy: 1.0000 - loss: 8.7652e-04 - val_accuracy: 1.0000 - val_loss: 0.0403
Epoch 5/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 91s 208ms/step - accuracy: 0.9952 - loss: 0.0173 - val_accuracy: 0.9332 - val_loss: 0.2763
Epoch 6/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 0s 74us/step - accuracy: 1.0000 - loss: 2.0878e-04 - val_accuracy: 0.9231 - val_loss: 0.1369
Epoch 7/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 96s 219ms/step - accuracy: 1.0000 - loss: 1.4928e-04 - val_accuracy: 0.9430 - val_loss: 0.2349
Epoch 8/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 0s 69us/step - accuracy: 1.0000 - loss: 7.5997e-06 - val_accuracy: 1.0000 - val_loss: 4.7003e-04
Epoch 9/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 99s 225ms/step - accuracy: 1.0000 - loss: 3.6063e-05 - val_accuracy: 0.9462 - val_loss: 0.2333
Epoch 10/10
440/440 ━━━━━━━━━━━━━━━━━━━━ 0s 81us/step - accuracy: 1.0000 - l

In [9]:
model.save('fruit_classifier_model.keras')

In [10]:
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_10 (Conv2D)              │ (None, 148, 148, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_10 (MaxPooling2D) │ (None, 74, 74, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_11 (Conv2D)              │ (None, 72, 72, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_11 (MaxPooling2D) │ (None, 36, 36, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_12 (Conv2D)              │ (None, 34, 34, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_12 (MaxPooling2D) │ (None, 17, 17, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_13 (Conv2D)              │ (None, 15, 15, 128)    │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_13 (MaxPooling2D) │ (None, 7, 7, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_3 (Flatten)             │ (None, 6272)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 512)            │     3,211,776 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 27)             │        13,851 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 10,399,379 (39.67 MB)

 Trainable params: 3,466,459 (13.22 MB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 6,932,920 (26.45 MB)

In [ ]:
test_dir = 'C:/Users/dmriv/Documents/GitHub/IAminibots/fruits-360/Test'


In [16]:
from tensorflow.keras.preprocessing.image import load_img, img_to_array
from tensorflow.keras.models import load_model
import numpy as np


model = load_model('fruit_classifier_model.keras')

def prepare_image(image_path):
    img = load_img(image_path, target_size=(150, 150))  
    img_array = img_to_array(img)  
    img_array = img_array / 255.0  
    img_array = np.expand_dims(img_array, axis=0)  
    return img_array

image_path = 'C:/Users/dmriv/Documents/GitHub/IAminibots/fruits-360/Test/Banana/14_100.jpg'  
prepared_img = prepare_image(image_path)

predictions = model.predict(prepared_img)

predicted_class = np.argmax(predictions, axis=1)

class_names = [
    'Apple Braeburn', 'Apple Crimson Snow', 'Apple Golden 1', 'Apple Golden 2',
    'Apple Golden 3', 'Apple Granny Smith', 'Apple Pink Lady', 'Apple Red 1',
    'Apple Red 2', 'Apple Red 3', 'Apple Red Delicious', 'Apple Red Yellow 1',
    'Apple Red Yellow 2', 'Banana', 'Banana Lady Finger', 'Banana Red',
    'Pear', 'Pear 2', 'Pear Abate', 'Pear Forelle', 'Pear Kaiser',
    'Pear Monster', 'Pear Red', 'Pear Stone', 'Pear Williams', 'Strawberry',
    'Strawberry Wedge'
]
 
predicted_class_name = class_names[predicted_class[0]]

print(f"Predicted class: {predicted_class_name}")


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step
Predicted class: Banana
